src/PackagePrivate/DatabaseTermIdsResolver.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use InvalidArgumentException;
+use stdClass;
+use Wikimedia\Rdbms\IDatabase;
+use Wikimedia\Rdbms\ILoadBalancer;
+use Wikimedia\Rdbms\IResultWrapper;
+
+/**
+ * Term ID resolver using the normalized database schema.
+ *
+ * @license GPL-2.0-or-later
+ */
+class DatabaseTermIdsResolver implements TermIdsResolver {
+
+	/** @var TypeIdsResolver */
+	private $typeIdsResolver;
+
+	/** @var ILoadBalancer */
+	private $lb;
+
+	/** @var IDatabase */
+	private $dbr = null;
+
+	/** @var IDatabase */
+	private $dbw = null;
+
+	/** @var string[] stash of data returned from the {@link TypeIdsResolver} */
+	private $typeNames = [];
+
+	public function __construct(
+		TypeIdsResolver $typeIdsResolver,
+		ILoadBalancer $lb
+	) {
+		$this->typeIdsResolver = $typeIdsResolver;
+		$this->lb = $lb;
+	}
+
+	/*
+	 * Term data is first read from the replica; if that returns less rows than we asked for,
+	 * then there are some new rows in the master that were not yet replicated, and we fall back
+	 * to the master. As the internal relations of the term store never change (for example,
+	 * a term_in_lang row will never suddenly point to a different text_in_lang), a master fallback
+	 * should never be necessary in any other case. However, callers need to consider where they
+	 * got the list of term IDs they pass into this method from: if it’s from a replica, they may
+	 * still see outdated data overall.
+	 */
+	public function resolveTermIds( array $termIds ): array {
+		$terms = [];
+		$this->connectDbr();
+
+		$replicaResult = $this->selectTerms( $this->dbr, $termIds );
+		$this->preloadTypes( $replicaResult );
+		$replicaTermIds = [];
+
+		foreach ( $replicaResult as $row ) {
+			$replicaTermIds[] = $row->wbtl_id;
+			$this->addResultTerms( $terms, $row );
+		}
+
+		if ( count( $replicaTermIds ) !== count( $termIds ) ) {
+			$masterTermIds = array_values( array_diff( $termIds, $replicaTermIds ) );
+			$this->connectDbw();
+			$masterResult = $this->selectTerms( $this->dbw, $masterTermIds );
+			$this->preloadTypes( $masterResult );
+			foreach ( $masterResult as $row ) {
+				$this->addResultTerms( $terms, $row );
+			}
+		}
+
+		return $terms;
+	}
+
+	private function selectTerms( IDatabase $db, array $termIds ): IResultWrapper {
+		return $db->select(
+			[ 'wbt_term_in_lang', 'wbt_text_in_lang', 'wbt_text' ],
+			[ 'wbtl_id', 'wbtl_type_id', 'wbxl_language', 'wbx_text' ],
+			[
+				'wbtl_id' => $termIds,
+				// join conditions
+				'wbtl_text_in_lang_id=wbxl_id',
```

Comment:
I think the example code I saw had no spaces around them, but I found regular JOINs to be very insufficiently documented in general :/ basically, the `IDatabase::select()` documentation says that for regular JOINs you don’t need `$join_conds`, and then doesn’t say what you should do instead – you’re just assumed to be familiar with what apparently are “ANSI-89” JOINs (`FROM x, y, z WHERE a = b`) rather than “ANSI-92” JOINs (`JOIN x ON a = b`).

But putting spaces around the `=`s certainly seems more consistent with MediaWiki’s general code style.